feat(core) : introduce AssumptionsWrapper

JSS95 opened the pull request:
Introduce AssumptionsWrapper

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

Please refer to the discussion in mailing list : https://groups.google.com/g/sympy/c/0NStDt45U-s

#### Other comments

`is_[...]` properties are given to `AssumptionsWrapper` by metaclass because instance construction was about 10 times faster than constructing without metaclass.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

- core
  - `is_eq()` can now take assumptions.

<!-- END RELEASE NOTES -->

JSS95 commented:
@asmeurer @oscarbenjamin 

Here is the blueprint for `AssumptionsWrapper` which was previously discussed in mailing list.

oscarbenjamin commented on sympy/core/assumptions.py:
```diff
@@ -573,3 +573,93 @@ def __init__(cls, *args, **kws):
+    def getit(self):
+        from sympy.assumptions import ask, Q
+        pred = getattr(Q, fact)
+        return ask(pred(self.expr), self.assumptions)
```
This `getit` function needs to be as fast as possible. I definitely wouldn't want the delayed import inside this function. It's possibly also faster to spell out the handlers explicitly like:
```
def is_positive(self):
    return ask(Q.positive(self), self.assumptions)
```

JSS95 replied:
Agreed. I moved `AssumptionsWrapper` from core to sympy.assumptions.ask module.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1428,7 +1442,7 @@ def split_real_imag(expr):
                         if rv is True:
                             rv = None
-        elif any(a.is_infinite for a in Add.make_args(n)):
+        elif any(AssumptionsWrapper(a, assumptions).is_infinite for a in Add.make_args(n)):
```
A one shot function would be better for cases like this. We don't need to make a wrapper instance just to call a single method:
```
def is_infinite(obj, assumptions):
    if assumptions is True:
         return obj.is_infinite
    else:
         return ask(Q.infinite(obj), assumptions)
```
I expect that to be faster and also to give tidier code when used.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1250,7 +1250,7 @@ def is_neq(lhs, rhs):
 
 
-def is_eq(lhs, rhs):
+def is_eq(lhs, rhs, assumptions=True):
```
I think that `assumptions=True` seems somewhat confusing here. I would expect that to mean something like "use the assumption system" when it fact it means "do not use any assumptions apart from the code assumptions".

JSS95 replied:
I made `None` as default assumption.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1408,16 +1420,24 @@ def split_real_imag(expr):
         if n2 is not None:
             return _sympify(n2 == 0)
 
+        def is_infinite(obj, assumptions):
+            # one shot function which is faster than AssumptionsWrapper
+            if assumptions is None:
+                return obj.is_infinite
+            return ask(Q.infinite(obj), assumptions)
```
I meant that functions like this could be defined in the same module as `AssumptionsWrapper` since they go together.

JSS95 replied:
I made assumptions/wrapper.py to define them.

oscarbenjamin commented:
Looks good but there should be some tests.

JSS95 commented:
@oscarbenjamin 
I added tests. But by the way, since we can use one-shot functions for every assumption keys, why do we need `AssumptionsWrapper`?

oscarbenjamin commented:
> But by the way, since we can use one-shot functions for every assumption keys, why do we need `AssumptionsWrapper`?

The advantage of `AssumptionsWrapper` is that if you make many queries on the same object when `assumptions=None` then apart from checking once that `assumptions=None` the other checks should be just as fast as they already are with the old assumptions.

It's also possible that by subclassing `Basic` the results of queries can be cached in `_assumptions` so that repeated queries are faster. Subclassing `Basic` might be slower though.

JSS95 commented:
Ok. Thanks for the explanation.

> It's also possible that by subclassing Basic the results of queries can be cached in _assumptions so that repeated queries are faster.

I think we can make the results cached without making AssumptionsWrapper Basic.

JSS95 commented:
@oscarbenjamin 
Making it `Basic` increased the construction time.

When subclassing `Basic`, 
```python
In [1]: from sympy.assumptions.wrapper import AssumptionsWrapper
In [2]: %timeit AssumptionsWrapper(x, Q.even(x))
2.15 µs ± 5.82 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
When not subclassing `Basic`:
```python
In [1]: from sympy.assumptions.wrapper import AssumptionsWrapper
In [2]: %timeit AssumptionsWrapper(x, Q.even(x))
1.96 µs ± 2.72 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

So I introduced caching without making it `Basic`.
previously:
```python
In [1]: from sympy.assumptions.wrapper import AssumptionsWrapper         
In [2]: a = x+y
In [3]: w = AssumptionsWrapper(a, ~Q.finite(x) & ~Q.finite(y) & Q.positive(x) & Q.positive(y))
In[4]: %timeit w.is_finite
5.4 ms ± 46.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
now:
```python
In [1]: from sympy.assumptions.wrapper import AssumptionsWrapper         
In [2]: a = x+y
In [3]: w = AssumptionsWrapper(a, ~Q.finite(x) & ~Q.finite(y) & Q.positive(x) & Q.positive(y))
In[4]: %timeit w.is_finite
112 ns ± 0.344 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

@asmeurer , would you review this PR? I think this is ready to be merged.

oscarbenjamin commented:
What I meant by cacheing is that the old assumptions system caches derived assumptions as well as specifically requested queries e.g.:
```
In [1]: x = Symbol('x', real=True)

In [2]: expr = 1 + x**2

In [3]: expr._assumptions
Out[3]: {}

In [4]: expr.is_positive
Out[4]: True

In [5]: expr._assumptions
Out[5]: 
{'positive': True,
 'finite': True,
 'infinite': False,
 'extended_positive': True,
 'extended_nonpositive': False,
 'nonpositive': False,
 'imaginary': False,
 'extended_real': True,
 'extended_nonzero': True,
 'extended_negative': False,
 'extended_nonnegative': True,
 'negative': False,
 'zero': False,
 'commutative': True,
 'real': True,
 'complex': True,
 'hermitian': True,
 'nonnegative': True,
 'nonzero': True}
```
Potentially that means that if you do:
```
e = 1 + x**2
_e = AssumptionsWrapper(e, Q.real(x))

if _e.is_positive:
    ...

if _e.is_real:
    ...
```
Then the `is_real` query comes from the cache rather than calling `ask` again. Maybe that's not needed though.

JSS95 commented:
I think that is a good idea. I added that change.

oscarbenjamin commented:
Looks good to me

JSS95 commented:
I will squash the commits and enable auto-merge, then.

JSS95 committed:
feat(core) : introduce AssumptionsWrapper

Introduce AssumptionsWrapper()
Introduce is_[...]() functions
Make is_eq able to take assumptions

oscarbenjamin commented:
The authors header was missing from the new 1.9 release notes page. I've added that in now.

Did you add the release note manually (it was already there when I looked)?

JSS95 commented:
@oscarbenjamin Yes, I did.
